Userland/Libraries/LibCrypto/Cipher/AES.cpp

```diff
@@ -152,24 +147,267 @@ void AESCipherKey::expand_encrypt_key(ReadonlyBytes user_key, size_t bits)
+{
+    AK::SIMD::i32x4 tb {};
+    AK::SIMD::i32x4 tc {};
+    AK::SIMD::i32x4 td {};
+
+    tb = AK::SIMD::as<AK::SIMD::i32x4>(b);
+    tc = AK::SIMD::as<AK::SIMD::i32x4>(__builtin_ia32_aeskeygenassist128(a, 0x00));
+    td = AK::SIMD::i32x4 { tc[2], tc[2], tc[2], tc[2] };
+    tc = AK::SIMD::i32x4 { 0, tb[0], tb[1], tb[2] };
+    tb ^= tc;
+    tc = AK::SIMD::i32x4 { 0, tc[0], tc[1], tc[2] };
+    tb ^= tc;
+    tc = AK::SIMD::i32x4 { 0, tc[0], tc[1], tc[2] };
+    tb ^= tc;
+    tb ^= td;
+    b = AK::SIMD::as<AK::SIMD::i64x2>(tb);
+}
+}
+#endif
+
+#if (ARCH(I386) || ARCH(X86_64))
+namespace
+{
+static void __attribute__((target("aes"), used)) AESCipherKey_expand_encrypt_key_256(u32* round_key, ReadonlyBytes user_key)
+{
+    auto const* user_ptr = user_key.data();
+    AK::SIMD::i64x2 ta {};
+    AK::SIMD::i64x2 tb {};
+    AK::SIMD::i64x2 tc {};
+
+    ta = AK::SIMD::load_unaligned<AK::SIMD::i64x2>(&user_ptr[0 * 16]);
+    tb = AK::SIMD::load_unaligned<AK::SIMD::i64x2>(&user_ptr[1 * 16]);
+    AK::SIMD::store_unaligned(&round_key[0 * 4], ta);
+    AK::SIMD::store_unaligned(&round_key[1 * 4], tb);
+    tc = __builtin_ia32_aeskeygenassist128(tb, 0x01); AESCipherKey_expand_encrypt_key_256_assist_a(ta, tc); AK::SIMD::store_unaligned(&round_key[ 2 * 4], ta); AESCipherKey_expand_encrypt_key_256_assist_b(ta, tb); AK::SIMD::store_unaligned(&round_key[ 3 * 4], tb);
+    tc = __builtin_ia32_aeskeygenassist128(tb, 0x02); AESCipherKey_expand_encrypt_key_256_assist_a(ta, tc); AK::SIMD::store_unaligned(&round_key[ 4 * 4], ta); AESCipherKey_expand_encrypt_key_256_assist_b(ta, tb); AK::SIMD::store_unaligned(&round_key[ 5 * 4], tb);
+    tc = __builtin_ia32_aeskeygenassist128(tb, 0x04); AESCipherKey_expand_encrypt_key_256_assist_a(ta, tc); AK::SIMD::store_unaligned(&round_key[ 6 * 4], ta); AESCipherKey_expand_encrypt_key_256_assist_b(ta, tb); AK::SIMD::store_unaligned(&round_key[ 7 * 4], tb);
+    tc = __builtin_ia32_aeskeygenassist128(tb, 0x08); AESCipherKey_expand_encrypt_key_256_assist_a(ta, tc); AK::SIMD::store_unaligned(&round_key[ 8 * 4], ta); AESCipherKey_expand_encrypt_key_256_assist_b(ta, tb); AK::SIMD::store_unaligned(&round_key[ 9 * 4], tb);
+    tc = __builtin_ia32_aeskeygenassist128(tb, 0x10); AESCipherKey_expand_encrypt_key_256_assist_a(ta, tc); AK::SIMD::store_unaligned(&round_key[10 * 4], ta); AESCipherKey_expand_encrypt_key_256_assist_b(ta, tb); AK::SIMD::store_unaligned(&round_key[11 * 4], tb);
+    tc = __builtin_ia32_aeskeygenassist128(tb, 0x20); AESCipherKey_expand_encrypt_key_256_assist_a(ta, tc); AK::SIMD::store_unaligned(&round_key[12 * 4], ta); AESCipherKey_expand_encrypt_key_256_assist_b(ta, tb); AK::SIMD::store_unaligned(&round_key[13 * 4], tb);
+    tc = __builtin_ia32_aeskeygenassist128(tb, 0x40); AESCipherKey_expand_encrypt_key_256_assist_a(ta, tc); AK::SIMD::store_unaligned(&round_key[14 * 4], ta);
```

Comment:
Last round does not get an `assist_b`?

Reply:
It works correctly, unit tests are passing, it is based on that white paper I linked above.